Last Item Reloading PR

TesterWoWDev opened the pull request:
idk what git is smoking today, but this really should be the last 1

TesterWoWDev committed:
version 1 custom item stats

TesterWoWDev committed:
Merge branch 'tswow:master' into master

TesterWoWDev committed:
replace mods that disapeared from git

TesterWoWDev committed:
damage function addition

TesterWoWDev committed:
fix the sql query

TesterWoWDev committed:
forgot to add this

ihm-tswow commented on tswow-core/Public/TSItemTemplate.h:
```diff
@@ -23,42 +23,56 @@ struct ItemTemplate;
     uint32 GetDisplayInfoID();
+    void SetDisplayInfoID(uint32 value);
     uint32 GetQuality();
+    void SetQuality(uint32 value);
```
Does it work to change these properties for normal item templates as well? What happens if they change? If the answer is ub/desyncs, they should probably not be on ItemTemplate itself but a dedicated subtype.

TesterWoWDev replied:
If modified it will modify the base item itself(unless using the copy function), as this was setup to allow for blizzlike item changes/reloads along with custom, it was left on the actual item_template

TesterWoWDev replied:
All of these functions could be moved to the virtual item on the other review message so it's only accessible from there, this just allows changing a blizzlike and reloading it to apply to all items.

ihm-tswow replied:
If that reload works reliably I don't mind this being available for normal item templates as well

TesterWoWDev replied:
I have not had an issue, but there is some live script boilerplate to make it work 'seamlessly'

TesterWoWDev replied:
tested spells and worked fine, so i added all of the fields and worst case, amend later.

TesterWoWDev committed:
Merge branch 'master' of https://github.com/tswow/tswow into tswow-master

TesterWoWDev committed:
Merge branch 'tswow-master'

TesterWoWDev committed:
trinity changes

TesterWoWDev committed:
merge latest

TesterWoWDev committed:
compiled tc

TesterWoWDev committed:
displayID stuff

TesterWoWDev committed:
Merge branch 'master' of https://github.com/tswow/tswow

TesterWoWDev committed:
i checked out the wrong branch on tc

TesterWoWDev committed:
fix settings merge

TesterWoWDev committed:
Merge branch 'master' of https://github.com/tswow/tswow

TesterWoWDev committed:
duplicate comment from merge

TesterWoWDev committed:
duplicate from merge

TesterWoWDev committed:
if trinity checks added to tsglobal

TesterWoWDev committed:
if trinity check added to TSItemTemplate

TesterWoWDev committed:
more trinity checks

TesterWoWDev committed:
more trinity checks

TesterWoWDev committed:
fix error logging

TesterWoWDev committed:
add if trinity to outfit methods

TesterWoWDev committed:
ifdefs

TesterWoWDev committed:
Merge branch 'master' of https://github.com/tswow/tswow

TesterWoWDev committed:
Merge branch 'tswow:tswow' into tswow-item-reload

TesterWoWDev committed:
fix lua header copy

ihm-tswow commented on tswow-core/Private/TSItemTemplate.cpp:
```diff
@@ -182,6 +234,190 @@ TSItemTemplate GetItemTemplate(uint32 entry)
+    TS_LOG_ERROR("tswow.api", "TSItemTemplate::SaveItemTemplate not implemented for AzerothCore");
+    #endif
+
+}
```
This should be a prepared statement, see `CharacterDatabase.h`/`CharacterDatabase.cpp`

TesterWoWDev replied:
Yeah, this was a quick put together as I think I want to remove some parts of this query as I don't think every field is required to be editable.

TesterWoWDev replied:
Also not sure if all parts are directly editable, from what I've tested they are but I haven't tested everything and I'm sure there must be some client limitations for some edits, just haven't come across them yet.

ihm-tswow replied:
i think that's good to verify for something like this so we don't publish stuff we have to remove. Of course, it can stay for now until someone wants to look into it

TesterWoWDev replied:
yeah, moving it to a prepared statement won't be hard once all of the parts are verified that they work. like I'm unsure if onUse or passive spells would effect with just a reload like this.

TesterWoWDev replied:
all fields are added but not all tested, waiting to move to prepared statements until all fields tested

ihm-tswow commented on tswow-core/Private/TSItem.cpp:
```diff
@@ -690,6 +690,24 @@ TSItemTemplate TSItem::GetTemplate()
+
+TSItemTemplate TSItem::GetTemplateCopy()
+{
+    return TSItemTemplate(new VirtualItemTemplate(item->GetTemplate()));
```
Memory leak? Who deletes the VirtualItemTemplate? Maybe it's suitable to handle this similarly to the dressnpcs addon, which uses smart pointers for dynamic data like that.

TesterWoWDev replied:
I grabbed this from Foe, it allows for a deep copy of the item but I am unsure if this would still be removed when the item itself would be deleted. Would need to look into that 1

ihm-tswow replied:
The constructor would still work if used as a smart pointer, but `TSItemTemplate` will not clean up the stray itemtemplate, so that's a memory leak. Normally, ItemTemplates are owned by the ObjectMgr since it's just static data, but with dynamic data we do need to consider where it should live, _or_ use a smart pointer. I think the latter is suitable here since it'd be used very similarly to TSOutfits/dynamic dressnpcs

ihm-tswow replied:
and i assume `Item` objects would need the ability to hold those smart pointers, so they stay loaded for the items they represent

TesterWoWDev replied:
currently all of the item_template data is just held in a hash map in the objmgr, in theory it only needs to be dynamic until it is allocated to the objmgr again. so using a template and then having the save function delete the object when it gets written to the objmgr

TesterWoWDev replied:
If I merge these 2 functions together, I could utilize the data and then delete the object. or just make SaveItemTemplate as should always be run last. unsure if there is a time where you would want the save and not the reload of the objmgr to apply values to the player.
SaveItemTemplate()
ReloadSingleItemTemplateObject(TSItemTemplate)

TesterWoWDev replied:
all of this kinda got redone in v2, so this conversation shouldnt matter as much. perhaps still a mem issue? but i think as it's only used and then cast away it's probably fine. It only allows this once per itemID anyways, so worst case it is a not horrible leak compared to before

ihm-tswow replied:
The general rule of thumb is that for anything you allocate with "new", you also need a way to "delete" it somewhere else. Storing things temporarily on the heap is fine, as long as it's deleted.

TesterWoWDev replied:
Well on the new system it is now handled by 
![image](https://user-images.githubusercontent.com/42619070/161059422-ecd5d3d5-3f8c-4925-9a60-220db90668b4.png)
and 
![image](https://user-images.githubusercontent.com/42619070/161059470-ed80b4d3-ef11-428c-a547-204b8516bdd8.png)

Where after it is called, it now goes over to the objmgr directly, where the core itself would handle it like a normal item_template so the serverside should delete it on it's own whenever it would be necessary

It still requires atleast 1 new as it is a seperate data that needs to be stored but this should still be safe as it allows the server to handle it on it's own without trying to keep track of any of the other pieces.

ihm-tswow replied:
This still leaks. You do `_itemTemplateStore[itemInfo->ItemId] = *itemInfo`, which creates a copy of your heap structure to store in the item template store, but does not track the heap pointer itself which is lost when that function goes out of scope. 

You don't really need to put anything on the heap here, just create an overload for `ObjectMgr::LoadSingleItemTemplateObject` to accept the old and new id, create a new VirtualItemTemplate (accept both the old object and new id in the constructor and call the super constructor / change entry from there instead of wrapping it in a TSItemTemplate for that) and store it directly inside the `_itemTemplateStore` by value (this avoids both the heap and unnecessary copies)

TesterWoWDev replied:
new commit fixing mem issues; [bfd444b](https://github.com/tswow/tswow/pull/472/commits/bfd444b9845bf07c6c7ee9f50f395abac2d1ff46)

TesterWoWDev committed:
Merge branch 'master' of https://github.com/tswow/tswow

TesterWoWDev committed:
update azc as well

TesterWoWDev committed:
delete past template before writing new

TesterWoWDev committed:
version 2

TesterWoWDev committed:
expose ItemTemplate attributes for reloading

TesterWoWDev committed:
fix return

TesterWoWDev committed:
fix misspells(again)

TesterWoWDev committed:
fix item reloading to work while equipped

TesterWoWDev committed:
remove unused functions

TesterWoWDev committed:
update CreateNewItemTemplate function declaration

TesterWoWDev committed:
i have no clue how this compiled the last time

TesterWoWDev committed:
Merge branch 'master' of https://github.com/tswow/tswow

TesterWoWDev committed:
rectify memory issues

ihm-tswow committed:
Merge branch 'master' into tswow-item-reload

ihm-tswow committed:
livescripts: fix GetExistingItemTemplate case

TesterWoWDev committed:
custom_item_template prepared statement

ihm-tswow committed:
livescripts: fix azerothcore fake loot data

ihm-tswow committed:
Merge branch 'master' into tswow-item-reload

ihm-tswow committed:
item-reload: remove VirtualItemTemplate

- not needed

ihm-tswow committed:
tswow: add tswow comments

ihm-tswow committed:
item-reload: rename custom ObjectMgr ItemTemplate methods

ihm-tswow committed:
item-reload: remove GetExistingItemTemplate

- not needed, we already have GetItemTemplate

ihm-tswow committed:
item-reload: rename LoadItemTemplateMutable -> GetItemTemplateMutable

ihm-tswow committed:
item-reload: avoid ItemTemplate const_cast on trinity

ihm-tswow committed:
item-reload: change SendItemQueryPacketWithTemplate to SendItemQueryPacket overload

ihm-tswow committed:
item-reload: load custom items on startup

ihm-tswow committed:
item-reload: remove LoadCustomItem livescript function

- should only happen on startup

ihm-tswow committed:
item-reload: rename CreateNewItemTemplate -> CreateItemTemplate

ihm-tswow committed:
item-reload: fix SendItemQueryPacket lua bindings

ihm-tswow committed:
item-reload: add cached item query support

ihm-tswow committed:
item-reload: add item.dbc disabler client patch

ihm-tswow committed:
item-reload: fix ItemTemplate::Save

- remove invalid spell index

ihm-tswow committed:
fix ItemTemplate::Save column types

ihm-tswow committed:
item-reload: fix InventoryType save column type

ihm-tswow committed:
item-reload: fix LoadCustomItemTemplates issues

- fix first row being skipped (while->do-while loop)

- remove unnecessary ItemTemplate heap allocation (store directly in _itemTemplateStore)

ihm-tswow committed:
item-reload: add ItemTemplate packet cache dirty flag

ihm-tswow committed:
azerothcore: add ItemTemplate::m_isDirty

TesterWoWDev committed:
add back SetSubClass

ihm-tswow committed:
item-reload fix CreateItemTemplate on azerothcore

ihm-tswow committed:
item-reload: fix stray submodule

ihm-tswow committed:
item-reload: fix overload error

tswow-admin commented on tswow-core/Public/TSGlobal.h:
```diff
@@ -20,6 +20,8 @@
 #include "TSItem.h"
 #include "TSBase.h"
 #include "TSString.h"
+TSItemTemplate TC_GAME_API CreateItemTemplate(uint32 entry,uint32 copyItemID = 38);
```
Why 38?

TesterWoWDev replied:
38 is a plain shirt incase people are unsure of what they want as a parent item and since there are so many minor fields it seemed easier to give it a parent even if a scripter doesnt want to. The real way it should be done is having the scripter create their own blank items in datascripts and using that for most usecases.

ihm-tswow replied:
addressed in the global.d.ts comments

tswow-admin commented on tswow-core/Public/global.d.ts:
```diff
@@ -1082,7 +1082,42 @@ declare interface TSPlayer extends TSUnit {
+     *
+     *
+     */
+    UpdateCache(): void
```
These are unlikely to be relevant outside of custom items, we should mention that in the comments.

TesterWoWDev replied:
Remove and Apply item mods is required for changing any stat, whether it blizzlike or custom, while it is equipped on the character. While true that generally you won't probably use this on a blizzlike item I wouldn't rule it out in however a server balances. Update cache is much the same but generally yes, while it is mostly for custom, there is a world where a server uses this to balance their blizzlike items as well.

ihm-tswow replied:
I think they meant it should be mentioned in the comments that these are functions to be used specifically when modifying item templates in general, so nobody tries messing with them for something like equips or something

ihm-tswow replied:
I wrote comments for this

tswow-admin commented on tswow-core/Public/global.d.ts:
```diff
@@ -8740,6 +8874,9 @@ declare function IsHolidayActive(holiday: uint16): boolean
 declare function StartGameEvent(event_id: uint16): void
 declare function StopGameEvent(event_id: uint16): void
+
+//defaults to 38
```
Change to proper C-style/jsdoc comment, explain what item this is

TesterWoWDev replied:
named the item + jsdoc style in latest commit [e6cc74e](https://github.com/tswow/tswow/pull/472/commits/e6cc74ec32ebff2fdb4dedb61cf4fcca8e35b356)

tswow-admin commented on tswow-scripts/util/ClientPatches.ts:
```diff
@@ -69,6 +70,67 @@ export function ClientPatches(
+                [0x1aa6e2,[0x90,0x90,0x90]],
+                [0x1aa821,[0x90,0x8d,0x4d,0x8,0xe8,0xa6,0xd9,0x1a]],
+                [0x1aa82a,[0x8b,0xf8,0xe9,0x67,0xbe,0x15,0x0]],
+                [0x1aa832,[0xc0]],
+                [0x1aa86c,[0x90,0x89,0xf8]],
+                [0x1aa8a2,[0x89,0xe1,0xe8,0xa7,0xd9,0x1a,0x0,0x83,0xc4,0x4]],
+                [0x1aa8b0,[0x90,0x90,0x90]],
+                [0x1aa9d0,[0x89,0xe1,0xe8,0x79,0xd8,0x1a,0x0,0x83,0xc4,0x4]],
+                [0x1aa9de,[0x90,0x90,0x90]],
+                [0x1aaafa,[0x89,0xe1,0xe8,0x4f,0xd7,0x1a,0x0,0x83,0xc4,0x4]],
+                [0x1aab08,[0x90,0x90,0x90]],
+                [0x1ab076,[0x89,0xe1,0xe8,0x53,0xd1,0x1a,0x0,0xe9,0x84,0xcc,0xff,0xff]],
+                [0x1ab083,[0xc0]],
+                [0x1ab0a9,[0x90,0x90,0x85,0xdb]],
+                [0x1ab316,[0x89,0xe1,0xe8,0x33,0xcf,0x1a,0x0,0x83,0xc4,0x4]],
+                [0x1ab324,[0x90,0x90,0x90]],
+                [0x306614,[0x8b,0x41,0x8,0x8d,0x48,0xc,0xe9,0x11,0x1b,0x5,0x0]],
+                [0x306620,[0xeb,0xf2,0xcc,0xcc,0xcc,0xcc]],
+                [0x306650,[0xeb,0x3a,0xcc,0xcc,0xcc,0xcc]],
+                [0x306683,[0x8d,0x48]],
+                [0x306686,[0xe9,0x45,0x1b,0x5,0x0,0xcc,0x8b,0x41,0x8,0x8d,0x48,0xc,0xe9,0xe9,0x1a,0x5,0x0,0xcc,0x8d,0x4d,0x8,0xe8,0xb0,0x1b,0x5]],
+                [0x3066a0,[0xe9,0x8c,0x41,0xea,0xff,0xcc,0xcc,0xcc,0xcc,0xcc,0xcc,0xcc,0xcc,0xcc]],
+                [0x306703,[0x8d,0x48]],
+                [0x306706,[0xe9,0x45,0x1b,0x5,0x0,0xcc,0xcc,0xcc,0xcc,0xcc,0xcc,0xcc,0xcc,0xcc,0xcc,0xcc,0xcc,0xcc,0xcc,0xcc,0xcc,0xcc,0xcc,0xcc,0xcc,0xcc,0xcc,0xcc,0xcc,0xcc,0xcc,0xcc,0xcc,0xcc,0xcc,0xcc,0xcc,0xcc,0xcc,0xcc]],
+                [0x306733,[0x8d,0x48]],
+                [0x306736,[0xe9,0x15,0x1c,0x5,0x0,0xcc,0xcc,0xcc,0xcc,0xcc,0xcc,0xcc,0xcc,0xcc,0xcc,0xcc,0xcc,0xcc,0xcc,0xcc,0xcc,0xcc,0xcc,0xcc,0xcc,0xcc,0xcc,0xcc,0xcc,0xcc,0xcc,0xcc,0xcc,0xcc,0xcc,0xcc,0xcc,0xcc,0xcc,0xcc]],
+                [0x309e03,[0x8d,0x48]],
+                [0x309e06,[0xe8,0x45,0xe4,0x4,0x0,0x90,0x90,0x90,0x90,0x90,0x90,0x90,0x90,0x90,0x90,0x90,0x90,0x90,0x90,0x90,0x90,0x90,0x90,0x90,0x90,0x90,0x90,0x90,0x90,0x90,0x90,0x90,0x90,0x90,0x90,0x90]],
+                [0x358136,[0x56,0x8b,0x31,0x89,0xf0]],
+                [0x35813c,[0x1,0x99,0x6a,0x0,0x68,0x70,0xeb,0x5e,0x0,0x33,0xc2,0x8d,0x4d,0xf8,0x51,0x2b,0xc2,0x50,0xb9,0x28,0xd8,0xc5,0x0,0xc7,0x45,0xf8]],
+                [0x358157,[0x0,0x0,0x0,0xc7,0x45,0xfc]],
+                [0x35815e,[0x0,0x0,0x0,0xe8,0xca,0x3c,0xf2,0xff,0x85,0xc0,0x74,0x8]],
+                [0x35816b,[0x40,0x4,0x5e,0x8b,0xe5,0x5d,0xc3,0x89,0xf0,0x5e,0x89,0xec,0x5d,0xe9,0xa9,0xe4,0xfa,0xff]],
+                [0x358186,[0x56,0x8b,0x31,0x89,0xf0]],
+                [0x35818c,[0x1,0x99,0x6a,0x0,0x68,0x70,0xeb,0x5e,0x0,0x33,0xc2,0x8d,0x4d,0xf8,0x51,0x2b,0xc2,0x50,0xb9,0x28,0xd8,0xc5,0x0,0xc7,0x45,0xf8]],
+                [0x3581a7,[0x0,0x0,0x0,0xc7,0x45,0xfc]],
+                [0x3581ae,[0x0,0x0,0x0,0xe8,0x7a,0x3c,0xf2,0xff,0x85,0xc0,0x74]],
+                [0x3581bb,[0x40,0x8,0x5e,0x8b,0xe5,0x5d,0xc3,0x89,0xf0,0x5e,0x89,0xec,0x5d,0xe9,0x89,0xe4,0xfa,0xff]],
+                [0x61be58,[0x7c,0x7c]]
+            ]),
```
This looks like a significant program, do we have permission to use it?

ihm-tswow replied:
https://model-changing.net/index.php?app=downloads&module=downloads&controller=view&id=314&do=findComment&comment=7592

>You are of course welcome to patch it by yourself﻿﻿

tswow-admin replied:
That's good enough for us

tswow-admin reviewed (changes requested):
Only minor points

TesterWoWDev committed:
item name + jsdoc

ihm-tswow committed:
item-reload: remove LoadSingleItemTemplate

ihm-tswow committed:
item-reload: remove leftover comment

ihm-tswow committed:
item-reload: rename custom_item_template.sql to match convention

ihm-tswow committed:
item-reload: improve CreateItemTemplate docs

ihm-tswow committed:
item-reload: improve item mod method descriptions

ihm-tswow committed:
item-reload: fix stray submodule (again)

ihm-tswow committed:
item-reload: fix equip comment

ihm-tswow commented:
Thanks for the PR!
